### ISSUE TITLE
Tracking issue towards MirageOS 3.0

### ISSUE THREAD

**Issue by amirmc** [mirage/mirage#592]
After chatting with @yomimono, this is an overall tracking issue to follow the progress to 3.0. Also see https://github.com/mirage/mirage/issues/571

### MirageOS Libs
- [x] Expose new targets `virtio` and `ukvm` via the [Solo5](https://github.com/Solo5) project — https://github.com/mirage/mirage/pull/565
- [x] Implement `is_unix` — _https://github.com/mirage/mirage/issues/575 and https://github.com/mirage/mirage/issues/578_
- [x] New Clock API (deprecate `CLOCK` module) —_https://github.com/mirage/mirage/pull/548 and https://github.com/mirage/mirage/pull/579_
- [x] Deprecate/remove functions using `base_context` (inc `--unix` and `--xen`) —_https://github.com/mirage/mirage/issues/560_
- [x] Add QubesOS as a target (i.e. `mirage configure -t qubes`)
- [x] Allow for specifying library version requirements in unikernel configuration file
- [x] Fix checksum behavior in mirage-tcpip (https://github.com/mirage/mirage-tcpip/pull/277)
- [x] Errors/result-type solution
- [x] MirageOS module type rename
- See pending list at https://github.com/mirage/mirage/issues/571

### Solo5
- [x] CI improvements and end to end testing
- [x] Stabilise virtio drivers
- [x] Audit ukvm interfaces
- See tracking issue at https://github.com/Solo5/solo5/issues/82

### Infrastructure(ish)
- [x] OPAM file desynchronization detector
- [x] List of deployable services — _See [comment](https://github.com/mirage/mirage/issues/592#issuecomment-253480558)_ and partially addressed in https://github.com/mirage/mirage-skeleton/commit/d136abff76256b7c257a539d4f7cdc9431930fd1
- [x] Improve/update/publish library docs — Tracked in #609 (see also #591)
- [x] Ensure Canopy builds with Mirage 3.0 — https://github.com/Engil/Canopy/pull/79 @samoht
- ~Performance harness. Tracked in #685~
- ~Changelogs — *Who? Or what tooling?*~

### Announcements
*For supporting blog posts on mirage.io, see https://github.com/mirage/mirage-www/issues/513*
- [x] Instructions/post/email on trying out the MirageOS beta — _@yomimono or @avsm_
- [x] Full documentation update for tutorials, wiki/install and other material on mirage-www -- _@yomimono_
- [x] Blog post on mirage.io about new features — _Draft/outline by @yomimono_ (available at https://github.com/yomimono/mirage-www/blob/mirageos-3-post/tmpl/blog/mirage-3.0-released.md) -- *Final URL: `https://mirage.io/blog/announcing-mirage-30-release`*
- [x] News release (involving Xen Project, Linux Foundation, and Docker) — _@amirmc_ (Ready)
- [x] Compose tweet for MirageOS twitter account — _@amirmc_
- [x] Docs on porting pre-existing MirageOS code to the shiny newness of 3.0 — _@yomimono_ ~_@avsm_~ -- *Final URL: `https://mirage.io/wiki/mirage2-to-mirage3`* see https://github.com/mirage/mirage-www/pull/516
- ~Updates to [Breaking Changes](https://mirage.io/wiki/breaking-changes) — _@yomimono?_~

### Release activity
- [x] Send pull requests for 3.0 package set to `opam-repository` — _@yomimono_ (https://github.com/ocaml/opam-repository/pull/8526)
- [x] Merge pull requests into `opam-repository` — _@avsm_
- [x] On [mirage/mirage-skeleton](https://github.com/mirage/mirage-skeleton), merge `mirage-dev` into `master` — _@yomimono_ (https://github.com/mirage/mirage-skeleton/pull/225)
- [x] On [mirage/mirage-www](https://github.com/mirage/mirage-www), merge `mirage-dev` into `master` (includes the 3.0 blog post & doc on porting code to 3.0) — _@avsm_ (https://github.com/mirage/mirage-www/pull/529) 
- [x] Send tweet from MirageOS twitter account — _@amirmc_
- [x] Send emails to MirageOS list and caml-list — _@yomimono_
- [x] GOTO Pub — ***Everyone***
- Check on build/deploy steps for `mirage-www` — _@avsm_ (Needs fixing: https://github.com/mirage/mirage-www/issues/531)

**Comment by amirmc** [comment]
@yomimono @mato @avsm (and others).

The above is an attempt to collate/link the various threads of work under one issue.  There's likely a lot missing from the above but I've tried to add the things that we are definitely planning to get done  and are working on (not just bouncing around as ideas).

If this works, then it'll be the main issue to review (periodically) to check overall progress. There are likely to be other tracking issues linked from this, as is already the case for Solo5.

**Comment by yomimono** [comment]
I've added the things I regard as must-do items to the MirageOS section, as well as a note on the need to do a look through mirage-www and identify sections which need updating.  mirage-www has a `mirage-dev` branch which builds against Mirage 3.0 and should talk about it by default.

**Comment by avsm** [comment]
See mirage/mirage#609 for the docs tracking

**Comment by hannesm** [comment]
After rethinking what I would like to have as a MirageOS3 release, I came to the conclusion that there are several parts missing (although worked on), namely: error handling.  From an operational view, I'd really like to have good examples (what is missing is log exfitration (via syslog) and statistics gathering (mrtg graphs)) of services which can be deployed by people:
- recursive dns resolver and dhcp for your home router (dns resolver for your laptop running as a virtual machine)
- dns server
- canopy / https servers -- let's encrypt integration
- firewall (qubes thingy)
- pop? imap? smtp? xmpp?

when mirage2 was a technology preview, and most people used them on their own laptops, I'd like mirage3 to be something people can deploy (on various targets, unix, xen, kvm, bhyve) and see whether it works out...

this means that it'll take some more time (depending on the number of people) to get towards a release, but the payoff may very well be that people will try it out (and run their secondary nameserver etc. on mirageos).

what do you think?

**Comment by yomimono** [comment]
I think these are really good points, and moreover developing these examples would be great projects for people who are interested in contributing to the release but don't want to wade into the interconnected thicket of libraries and module types involved in contributing to the API changes.

**Comment by mato** [comment]
Regarding improving our examples: `mirage-skeleton` is currently a "grab-bag of all things", some very low-level (eg. vchan stuff), some full high-level applications (eg. static_website). So, a curated repository with _documented_ examples of high-level applications is definitely something we lack at the moment. I would start with the parts we have now:
* Recursive DNS resolver
* DNS server
* Canopy/HTTPS
* Qubes Firewall (maybe, this is pretty niche)

Regarding LE integration: There are two options here:

1. Integrate LE at _build_ time. Depend on [Certbot](https://certbot.eff.org/) (ex `letsencrypt`) using `standalone` mode, bake the resulting secrets into the unikernel image. To renew you re-run the build process and redeploy.
2. Integrate LE at _run_ time. This implies secrets are persisted outside the unikernel image (i.e. on a block device or some other persistent store the unikernel has trusted access to). We don't have a standardised, cross-backend solution for this at the moment.

I presume you meant the second option?

Regarding logs and metrics: How about channeling everything through `mirage-log`, including metrics (eg. TCP and GC stats)? To get the data out we then have two (cross-backend) options:

1. Pipe the console to something on the host. Simplest option, should work everywhere.
2. Get the data out via remote `syslog`. From an ops perspective, this probably requires two network interfaces (frontend and backend). Note that Solo5 backends only support a single network interface at present.

**Comment by amirmc** [comment]
I second the comments about well documented, usable, high-level apps.

However, something I'm wary of is adding things to a backlog of items (since that's always easy) versus defining the minimal set of things needed for the release (perhaps more difficult as it involves saying "no, not right now").  That doesn't mean we should _stop_ at the minimal set, just that we should maintain clarity over what those things are so that we can chart progress.

For example, I'd like to restructure the top comment here to reflect the overall goals a little better but it's not clear (to me) where the lines really are. e.g. Delpoying a Solo5 backend for mirage.io involves a bunch of necessary stuff and perhaps some nice-to- have stuff.

**Comment by samoht** [comment]
Can I have a new shiny conduit library for MirageOS3 too? :-)

**Comment by yomimono** [comment]
@samoht: I suggest you ask over on [conduit](https://github.com/mirage/ocaml-conduit), unless there are API dependencies for your request ;)

**Comment by samoht** [comment]
done: https://github.com/mirage/ocaml-conduit/issues/157

**Comment by hannesm** [comment]
@amirmc what I wanted to express is the other side of a release.  We can cut a release right now, and claim it to be 3.0, but what do we get from it?  By having several nice useful applications people can play around with, I've the hope we'll get more people who read the announcement and then stick to try out something, report bugs, get involved.

@mato for me, mirage-skeleton feels like a laundry list with partially too complex examples.  Best to be used in presentations since the code is short enough to fit on a slide.

@mato exfiltration via console scraping is not my favourite mechanism.  We can use syslog (over TCP/UDP/TLS over TCP -- no requirement for a second network device) in an encrypted and authenticated fashion -- missing piece is a logs reporter which emits syslog messages (there is a syslog library).  I agree that statistics can be exfiltrated the same way and fed into common utilities (mrtg/...).

@mato LE integration - I thought about embedding the long-term secrets in a unikernel and let itself after 90 days of sleep renew its certificates.  There's some preliminary implementation of ACME, which we should test and extend further.  We do have some mirage-block devices, and could as well use a mirage-block-ccm on top of that, and then a ocaml-fat file system?  I haven't looked too deep into that, but having a way to write data persistently is surely something we want to provide as a feature.

@samoht surely you can have it.  You're in charge :) I'm still not sure what the properties of conduits should be (esp. regarding authenticity)

**Comment by mato** [comment]
> mirage-skeleton feels like a laundry list with partially too complex examples. Best to be used in presentations since the code is short enough to fit on a slide.

I think we're in agreement here (also with what @amirmc is saying).

> exfiltration via console scraping is not my favourite mechanism. We can use syslog (over TCP/UDP/TLS over TCP -- no requirement for a second network device) in an encrypted and authenticated fashion

IMO sending logs out over an _internet-connected_ interface _implies_ encrypt+auth. I'm not aware of a protocol for encrypted `syslog`? In the absence of encrypt+auth sending the data out over a separate interface is a requirement.

If your single interface is _not_ internet-connected (eg. NAT or Docker) then plain old syslog over UDP is fine.

> I thought about embedding the long-term secrets in a unikernel and let itself after 90 days of sleep renew its certificates.

Without persisting those secrets somewhere you'll lose the ACME registration data anytime the unikernel is redeployed/restarted.

**Comment by hannesm** [comment]
@mato TLS for syslog is defined in RFC 5425 https://tools.ietf.org/html/rfc5425 -- should be straightforward

@mato yes, persistent writable data would be great to have.  so far I haven't seen it working in MirageOS.  Crunch is a RO volume, Irmin/git does remote fetch (but I haven't seen any push).

**Comment by hannesm** [comment]
some news on [syslog](https://hannesm.github.io/logs-syslog/doc/) - implemented and released to opam (for Mirage2) https://github.com/ocaml/opam-repository/pull/7720, and in mirage-dev for Mirage3 (UDP/TCP/TLS support).

Interop with syslog-ng works via mutually authenticated Canopy (see https://github.com/hannesm/Canopy/commit/732f3245ebbcf5f79e3d47ae797ca778568f896b) (I run it live on my blog), but no integration into the mirage tool yet.

**Comment by avsm** [comment]
Performance harness tracking issue now at: https://github.com/mirage/mirage/issues/685

**Comment by hannesm** [comment]
for the tagging vs opam remote story: IMHO we should make proper (tarball) releases, so they have real version numbers (watermarks), then provide a mirage-dev branch (named `#beta`, together with mirage-www `#beta` and mirage-skeleton `#beta`), and as well a shell script which calls for each library out to `opam pin add <foo> <https://where/my/tarball/is/3.0.0~beta1.tar>`... pins are so much less intrusive (in contrast to opam remote)

we can then continue to develop/bugfix on the mirage-dev branches.

certainly the conflicts/upper version bounds in the main opam-repository can already be added (maybe this is even scriptable, now that opam-file-format is around)!?

**Comment by dbuenzli** [comment]
> for the tagging vs opam remote story: IMHO we should make proper (tarball) releases, so they have real version numbers (watermarks), then provide a mirage-dev branch (named #beta, together with mirage-www #beta and mirage-skeleton #beta),

If you use pins there's no need to make release tarballs you can simply make an annotated tag `v3.0.0~beta` and make `git-pin`s on these tags (this also allows you to move the tag). Pins will be correctly watermarked (they won't in an OPAM remote due to the inconclusive discussion [here](https://github.com/mirage/functoria/issues/67)).

**Comment by amirmc** [comment]
Shuffled some of the items in the main list and added an 'Infrastructure(ish)' section (for things that aren't quite library specific).

I also added an item regarding a notice about how users get onto the beta (by whatever mechanism is decided — some discussion in the [last IRC meeting](http://canopy.mirage.io/irclogs/16-11-2016)).

**Comment by amirmc** [comment]
Added an item about changelogs.

**Comment by hannesm** [comment]
(updated link to Canopy MirageOS3 PR)

**Comment by yomimono** [comment]
I consider "Full documentation update for tutorials, wiki/install and other material on mirage-www -- @yomimono" to be complete, but second opinions are very welcome.  Please see the `mirage-dev` branches of `mirage-www` and `mirage-skeleton` to form them. :)

**Comment by yomimono** [comment]
re: changelogs - what's appropriate to collate and surface?  The summary at https://github.com/mirage/mirage/releases/tag/v3.0.0 has a human-intermediated collection of changes in non-Mirage libraries (particularly those that used to be in the `mirage/mirage` repo but departed the nest in this release).

As part of tagging the versions of libraries involved in MirageOS 3 I generally made at least the most recent versions of each changelog automatically readable by `topkg`, so concatenating all of the release data into one file isn't difficult, but I'm also not sure it's useful.

**Comment by yomimono** [comment]
A draft blog post announcing the release is available at https://github.com/yomimono/mirage-www/blob/mirageos-3-post/tmpl/blog/mirage-3.0-released.md  .

**Comment by talex5** [comment]
The blog post looks pretty good. But as it's likely to get linked from various new sites, might it be worth adding a paragraph at the top saying what Mirage is? e.g. something along the lines of:

> We're excited to announce MirageOS 3.0! MirageOS is a modern, modular unikernel operating system that allows the creation of small, secure, legacy-free services. Mirage applications can be compiled to run as self-contained virtual machines (a few MB in size) on Xen or KVM hosts, or as regular Unix processes (allowing access to regular debugging tools).

Likewise, I think the Solo5 section needs a line or two saying what it is.

**Comment by hannesm** [comment]
- the link to docs.mirage.io is using https, but the server does not do https.
- I'd remove the names from the text, and only have names in the bullet list
- the 80000 loc less is a nice number, but more useful expressed in relation with the total number of lines (or N percent)
- I agree with @talex5 above, and would add a link to `getting started` (e.g. the revised tutorial, some example unikernels) so that the enthusiastic reader can jump into running their unikernels :) (see also https://github.com/mirage/mirage/issues/592#issuecomment-253498598 https://github.com/mirage/mirage/issues/592#issuecomment-253480558 above)

**Comment by avsm** [comment]
I'll deploy docs.mirage.io to use https via LE next week -- travelling this week.

**Comment by yomimono** [comment]
Thanks for the comments!  @talex5, I liked your paragraph and put it in as-is.  @hannesm, I think I've addressed everything but your third point; I'm still not really happy with how this is surfaced either.  It might be easier to figure out what to write here when we've thought more about what to write in a larger article.

**Comment by amirmc** [comment]
@yomimono @avsm, I need to add the URL for the blog post to the news release.  Since I haven't heard other comments about this, I'm using `https://mirage.io/blog/announcing-mirage-30-release` (which is similar in form to previous releases).  Hope that's ok.

**Comment by yomimono** [comment]
The URL above is fine :) Worst case, we can add a redirect to the unikernel that runs the blog, but that's definitely a reasonable URL in the existing web server and I think it'll work for @avsm's rewrite as well.

**Comment by amirmc** [comment]
@yomimono, thanks!  I also need the URL for the (upcoming) guide on moving from Mirage 2 to Mirage 3.  I've looked at the structure of the pages and, I'll use `https://mirage.io/wiki/mirage2-to-mirage3` for that content.  The link will be in the news announcement.

**Comment by yomimono** [comment]
Thanks @amirmc, that link is great.  I'm hoping to PR that yet today (UTC).

**Comment by samoht** [comment]
I have updated the first post to add a link to https://github.com/mirage/mirage-www/pull/514

**Comment by amirmc** [comment]
Hi folks, this issue and https://github.com/mirage/mirage-www/issues/513 were getting confusing.  

This issue (https://github.com/mirage/mirage/issues/592) was intended to track the _minimal_ set of items required for the release and announcement.  To clarify things, I've removed some items from here and copied them over to https://github.com/mirage/mirage-www/issues/513.

TL;DR. If it's required for the announcement, it should be here. Otherwise, I moved it.

**Comment by mato** [comment]
@yomimono Should there be a checklist item on this issue to "Merge the `mirage-dev` branch of mirage/mirage-skeleton into `master`?

**Comment by amirmc** [comment]
@mato, that's a good point.  For the sake of tracking that somewhere, I've added the following two items to the checklist.

- On [mirage/mirage-skeleton](https://github.com/mirage/mirage-skeleton), merge `mirage-dev`->`master` — _@yomimono_
- On [mirage/mirage-www](https://github.com/mirage/mirage-www), merge `mirage-dev`->`master` — _@yomimono_

**Comment by yomimono** [comment]
There is an additional set of post-release steps that I've been keeping in my head, but can record elsewhere - e.g. remove `EXTRA_REMOTES=https://github.com/mirage/mirage-dev.git` from all of the repositories, retire mirageos-3-beta, etc.

**Comment by amirmc** [comment]
@yomimono It'd be good to have those out of your head and recorded somewhere as a checklist.  This issue is fine or feel free to make another.  It would allow someone else to keep up in case your internets get sand in them. :-P

**Comment by yomimono** [comment]
@amirmc I'll pull that together once I'm done with the promised porting guide.

**Comment by amirmc** [comment]
I've tidied up the top post and added a subsection for release activity. Copied here for those who like the email notifications. Please add anything you think I missed.

### Release activity
- [ ] Send pull requests for 3.0 package set to `opam-repository` — _@yomimono_
- [ ] Merge pull requests into `opam-repository` — ***avsm/yallop/dsheets?***
- [ ] On [mirage/mirage-skeleton](https://github.com/mirage/mirage-skeleton), merge `mirage-dev` into `master` — _@yomimono_
- [ ] On [mirage/mirage-www](https://github.com/mirage/mirage-www), merge `mirage-dev` into `master` (includes the 3.0 blog post) — _@yomimono_ 
- [ ] Check on deploy for `mirage-www` — _@mor1_
- [ ] Send tweet from MirageOS twitter account — _@amirmc_
- [ ] GOTO Pub — ***Everyone***

**Comment by avsm** [comment]
I'll merge into opam-repository -- will need to check the revdeps on the PR with datakit

**Comment by hannesm** [comment]
..there's one more thing:  can we link to http://docs.mirage.io (and/or get https there) from a prominent place on https://mirage.io (IIRC we used to have a rider on the top with `API docs`, maybe reactivate that)?

**Comment by yomimono** [comment]
https://github.com/mirage/mirage-www/pull/461 was the commit that removed the API link, for reference.

**Comment by yomimono** [comment]
I recently (yes, really) joined [caml-list](https://sympa.inria.fr/sympa/arc/caml-list) and sometimes people announce their software releases there.  Maybe we should do that too.

**Comment by yomimono** [comment]
Also, we need to change the /topic in `#mirage` on FreeNode.  I think @avsm knows a bot that has ops there.

**Comment by yomimono** [comment]
I believe all of the items here are done, with the exception of the last one which will shortly be in progress.  The items at https://github.com/mirage/mirage/issues/792 have yet to be tackled.

**Comment by amirmc** [comment]
Waiting to hear that the build/deploy steps are all still as expected. :)

**Comment by amirmc** [comment]
The build/deploy scripts need updating, so I've open https://github.com/mirage/mirage-www/issues/531 to track that.  Since everything else is done, I'm closing this issue.

Well done everyone! 🎉 😄
